File: plugin/src/knowledge/providers/defillama.ts
import { requestUrl } from "obsidian";
import type { KnowledgeProvider, LookupOptions, LookupResult, KnowledgeEntry } from "../types";

const DEFILLAMA_API = "https://api.llama.fi";

interface DefiProtocol {
  id: string;
  name: string;
  slug: string;
  symbol?: string;
  url?: string;
  description?: string;
  chain?: string;
  chains?: string[];
  tvl?: number;
  chainTvls?: Record<string, number>;
  change_1h?: number;
  change_1d?: number;
  change_7d?: number;
  category?: string;
  logo?: string;
  twitter?: string;
  gecko_id?: string;
  mcap?: number;
}

interface DefiChain {
  name: string;
  tvl: number;
  chainId?: number;
  gecko_id?: string;
  tokenSymbol?: string;
}

export class DefiLlamaProvider implements KnowledgeProvider {
  readonly name = "defillama";

  async isAvailable(): Promise<boolean> {
    try {
      const response = await requestUrl({
        url: `${DEFILLAMA_API}/protocols`,
        method: "GET",
        throw: false,
      });
      return response.status === 200;
    } catch {
      return false;
    }
  }

  async lookup(query: string, options?: LookupOptions): Promise<LookupResult> {
    const maxResults = options?.maxResults ?? 5;

    try {
      // Check if query is a chain name
      const chainEntry = await this.lookupChain(query);
      if (chainEntry) {
        return {
          success: true,
          provider: this.name,
          entries: [chainEntry],
        };
      }

      // Search protocols
      const entries = await this.searchProtocols(query, maxResults);

      return {
        success: true,
        provider: this.name,
        entries,
      };
    } catch (error) {
      const message = error instanceof Error ? error.message : "Unknown error";
      return {
        success: false,
        provider: this.name,
        entries: [],
        error: `DefiLlama lookup failed: ${message}`,
      };
    }
  }

  private async searchProtocols(query: string, limit: number): Promise<KnowledgeEntry[]> {
    const response = await requestUrl({
      url: `${DEFILLAMA_API}/protocols`,
      method: "GET",
      throw: false,
    });

    if (response.status !== 200) {
      throw new Error(`Failed to fetch protocols: ${response.status}`);
    }

    const protocols = response.json as DefiProtocol[];
    const queryLower = query.toLowerCase();

    // Filter and score matches
    const matches = protocols
      .filter((p) => {
        const nameMatch = p.name.toLowerCase().includes(queryLower);
        const symbolMatch = p.symbol?.toLowerCase().includes(queryLower);
        const categoryMatch = p.category?.toLowerCase().includes(queryLower);
        return nameMatch || symbolMatch || categoryMatch;
      })
      .sort((a, b) => {
        // Exact name match first
        const aExact = a.name.toLowerCase() === queryLower ? 1 : 0;
        const bExact = b.name.toLowerCase() === queryLower ? 1 : 0;
        if (aExact !== bExact) return bExact - aExact;

        // Then by TVL
        return (b.tvl ?? 0) - (a.tvl ?? 0);
      })
      .slice(0, limit);

    return matches.map((protocol) => this.protocolToEntry(protocol));
  }

  private async lookupChain(name: string): Promise<KnowledgeEntry | null> {
    const response = await requestUrl({
      url: `${DEFILLAMA_API}/v2/chains`,
      method: "GET",
      throw: false,
    });

    if (response.status !== 200) {
      return null;
    }

    const chains = response.json as DefiChain[];
    const nameLower = name.toLowerCase();

    const chain = chains.find(
      (c) => c.name.toLowerCase() === nameLower
    );

    if (!chain) {
      return null;
    }

    const tvlFormatted = this.formatTvl(chain.tvl);

    return {
      title: chain.name,
      summary: `Blockchain with ${tvlFormatted} TVL`,
      url: `https://defillama.com/chain/${chain.name}`,
      source: "defillama",
      metadata: {
        type: "chain",
        tvl: chain.tvl,
        tvlFormatted,
        chainId: chain.chainId,
        geckoId: chain.gecko_id,
        tokenSymbol: chain.tokenSymbol,
      },
    };
  }

  private protocolToEntry(protocol: DefiProtocol): KnowledgeEntry {
    const lines: string[] = [];

    if (protocol.description) {
      lines.push(protocol.description.slice(0, 200));
    }

    const tvl = this.formatTvl(protocol.tvl);
    lines.push(`TVL: ${tvl}`);

    if (protocol.category) {
      lines.push(`Category: ${protocol.category}`);
    }

    if (protocol.chains && protocol.chains.length > 0) {
      lines.push(`Chains: ${protocol.chains.slice(0, 5).join(", ")}`);
    }

    if (protocol.change_1d !== undefined) {
      const changeSign = protocol.change_1d >= 0 ? "+" : "";
      lines.push(`24h Change: ${changeSign}${protocol.change_1d.toFixed(2)}%`);
    }

    return {
      title: protocol.name,
      summary: lines.join("\n"),
      url: `https://defillama.com/protocol/${protocol.slug}`,
      source: "defillama",
      metadata: {
        type: "protocol",
        id: protocol.id,
        slug: protocol.slug,
        symbol: protocol.symbol,
        category: protocol.category,
        tvl: protocol.tvl,
        tvlFormatted: tvl,
        chains: protocol.chains,
        change1h: protocol.change_1h,
        change1d: protocol.change_1d,
        change7d: protocol.change_7d,
        mcap: protocol.mcap,
        twitter: protocol.twitter,
        geckoId: protocol.gecko_id,
        logo: protocol.logo,
        website: protocol.url,
      },
    };
  }

  private formatTvl(tvl?: number): string {
    if (!tvl) return "N/A";

    if (tvl >= 1e9) {
      return `$${(tvl / 1e9).toFixed(2)}B`;
    } else if (tvl >= 1e6) {
      return `$${(tvl / 1e6).toFixed(2)}M`;
    } else if (tvl >= 1e3) {
      return `$${(tvl / 1e3).toFixed(2)}K`;
    }
    return `$${tvl.toFixed(2)}`;
  }

  async getProtocolTvlHistory(slug: string): Promise<Array<{ date: number; tvl: number }> | null> {
    const response = await requestUrl({
      url: `${DEFILLAMA_API}/protocol/${slug}`,
      method: "GET",
      throw: false,
    });

    if (response.status !== 200) {
      return null;
    }

    const data = response.json as { tvl?: Array<{ date: number; totalLiquidityUSD: number }> };

    if (!data.tvl) {
      return null;
    }

    return data.tvl.map((point) => ({
      date: point.date,
      tvl: point.totalLiquidityUSD,
    }));
  }
}

File: crates/mcp/src/knowledge/defillama.rs
use reqwest::blocking::Client;
use serde::Deserialize;
use std::collections::HashMap;

use super::{KnowledgeEntry, KnowledgeProvider, LookupOptions, LookupResult};

const DEFILLAMA_API: &str = "https://api.llama.fi";

pub struct DefiLlamaProvider {
    client: Client,
}

#[derive(Debug, Deserialize)]
#[allow(dead_code)]
struct Protocol {
    id: String,
    name: String,
    slug: String,
    symbol: Option<String>,
    url: Option<String>,
    description: Option<String>,
    chain: Option<String>,
    chains: Option<Vec<String>>,
    tvl: Option<f64>,
    change_1h: Option<f64>,
    change_1d: Option<f64>,
    change_7d: Option<f64>,
    category: Option<String>,
    logo: Option<String>,
    twitter: Option<String>,
    gecko_id: Option<String>,
    mcap: Option<f64>,
}

#[derive(Debug, Deserialize)]
#[allow(dead_code)]
struct Chain {
    name: String,
    tvl: f64,
    #[serde(rename = "chainId")]
    chain_id: Option<i64>,
    gecko_id: Option<String>,
    #[serde(rename = "tokenSymbol")]
    token_symbol: Option<String>,
}

impl DefiLlamaProvider {
    pub fn new() -> Self {
        Self {
            client: Client::builder()
                .user_agent("vault-tree-mcp/0.1 (https://github.com/Hydepwns/vault-tree)")
                .build()
                .unwrap_or_else(|_| Client::new()),
        }
    }

    fn format_tvl(tvl: Option<f64>) -> String {
        match tvl {
            None => "N/A".to_string(),
            Some(v) if v >= 1e9 => format!("${:.2}B", v / 1e9),
            Some(v) if v >= 1e6 => format!("${:.2}M", v / 1e6),
            Some(v) if v >= 1e3 => format!("${:.2}K", v / 1e3),
            Some(v) => format!("${:.2}", v),
        }
    }

    fn protocol_to_entry(&self, protocol: &Protocol) -> KnowledgeEntry {
        let mut lines = Vec::new();

        if let Some(desc) = &protocol.description {
            lines.push(desc.chars().take(200).collect::<String>());
        }

        let tvl = Self::format_tvl(protocol.tvl);
        lines.push(format!("TVL: {}", tvl));

        if let Some(cat) = &protocol.category {
            lines.push(format!("Category: {}", cat));
        }

        if let Some(chains) = &protocol.chains {
            if !chains.is_empty() {
                let chain_list: Vec<_> = chains.iter().take(5).cloned().collect();
                lines.push(format!("Chains: {}", chain_list.join(", ")));
            }
        }

        if let Some(change) = protocol.change_1d {
            let sign = if change >= 0.0 { "+" } else { "" };
            lines.push(format!("24h Change: {}{:.2}%", sign, change));
        }

        let mut metadata = HashMap::new();
        metadata.insert("type".to_string(), serde_json::json!("protocol"));
        metadata.insert("id".to_string(), serde_json::json!(protocol.id));
        metadata.insert("slug".to_string(), serde_json::json!(protocol.slug));
        if let Some(sym) = &protocol.symbol {
            metadata.insert("symbol".to_string(), serde_json::json!(sym));
        }
        if let Some(cat) = &protocol.category {
            metadata.insert("category".to_string(), serde_json::json!(cat));
        }
        if let Some(tvl) = protocol.tvl {
            metadata.insert("tvl".to_string(), serde_json::json!(tvl));
        }
        if let Some(chains) = &protocol.chains {
            metadata.insert("chains".to_string(), serde_json::json!(chains));
        }

        KnowledgeEntry {
            title: protocol.name.clone(),
            summary: lines.join("\n"),
            url: Some(format!("https://defillama.com/protocol/{}", protocol.slug)),
            source: "defillama".to_string(),
            metadata: Some(metadata),
        }
    }

    fn search_protocols(&self, query: &str, limit: usize) -> Result<Vec<KnowledgeEntry>, String> {
        let url = format!("{}/protocols", DEFILLAMA_API);

        let response = self.client.get(&url).send().map_err(|e| e.to_string())?;

        if !response.status().is_success() {
            return Err(format!("failed to fetch protocols: {}", response.status()));
        }

        let protocols: Vec<Protocol> = response.json().map_err(|e| e.to_string())?;

        let query_lower = query.to_lowercase();

        let mut matches: Vec<_> = protocols
            .iter()
            .filter(|p| {
                let name_match = p.name.to_lowercase().contains(&query_lower);
                let symbol_match = p.symbol.as_ref().map(|s| s.to_lowercase().contains(&query_lower)).unwrap_or(false);
                let category_match = p.category.as_ref().map(|c| c.to_lowercase().contains(&query_lower)).unwrap_or(false);
                name_match || symbol_match || category_match
            })
            .collect();

        matches.sort_by(|a, b| {
            let a_exact = a.name.to_lowercase() == query_lower;
            let b_exact = b.name.to_lowercase() == query_lower;
            if a_exact != b_exact {
                return b_exact.cmp(&a_exact);
            }
            b.tvl.partial_cmp(&a.tvl).unwrap_or(std::cmp::Ordering::Equal)
        });

        Ok(matches
            .into_iter()
            .take(limit)
            .map(|p| self.protocol_to_entry(p))
            .collect())
    }

    fn lookup_chain(&self, name: &str) -> Result<Option<KnowledgeEntry>, String> {
        let url = format!("{}/v2/chains", DEFILLAMA_API);

        let response = self.client.get(&url).send().map_err(|e| e.to_string())?;

        if !response.status().is_success() {
            return Ok(None);
        }

        let chains: Vec<Chain> = response.json().map_err(|e| e.to_string())?;

        let name_lower = name.to_lowercase();
        let chain = chains.iter().find(|c| c.name.to_lowercase() == name_lower);

        match chain {
            None => Ok(None),
            Some(c) => {
                let tvl = Self::format_tvl(Some(c.tvl));

                let mut metadata = HashMap::new();
                metadata.insert("type".to_string(), serde_json::json!("chain"));
                metadata.insert("tvl".to_string(), serde_json::json!(c.tvl));
                metadata.insert("tvlFormatted".to_string(), serde_json::json!(tvl));
                if let Some(id) = c.chain_id {
                    metadata.insert("chainId".to_string(), serde_json::json!(id));
                }
                if let Some(sym) = &c.token_symbol {
                    metadata.insert("tokenSymbol".to_string(), serde_json::json!(sym));
                }

                Ok(Some(KnowledgeEntry {
                    title: c.name.clone(),
                    summary: format!("Blockchain with {} TVL", tvl),
                    url: Some(format!("https://defillama.com/chain/{}", c.name)),
                    source: "defillama".to_string(),
                    metadata: Some(metadata),
                }))
            }
        }
    }
}

impl Default for DefiLlamaProvider {
    fn default() -> Self {
        Self::new()
    }
}

impl KnowledgeProvider for DefiLlamaProvider {
    fn name(&self) -> &'static str {
        "defillama"
    }

    fn is_available(&self) -> bool {
        self.client
            .get(format!("{}/protocols", DEFILLAMA_API))
            .send()
            .map(|r| r.status().is_success())
            .unwrap_or(false)
    }

    fn lookup(&self, query: &str, options: &LookupOptions) -> LookupResult {
        let limit = options.max_results.unwrap_or(5);

        // Check if query is a chain name
        if let Ok(Some(entry)) = self.lookup_chain(query) {
            return LookupResult::success(self.name(), vec![entry]);
        }

        // Search protocols
        match self.search_protocols(query, limit) {
            Ok(entries) => LookupResult::success(self.name(), entries),
            Err(e) => LookupResult::error(self.name(), e),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    #[ignore] // Requires network
    fn defillama_lookup() {
        let provider = DefiLlamaProvider::new();
        let result = provider.lookup("uniswap", &LookupOptions::default());
        assert!(result.success);
    }
}
